File: 399-evaluate-division.cc
#include <string>
#include <unordered_map>
#include <vector>

using namespace std;

class Solution {
public:
    vector<double> calcEquation(vector<vector<string>>& equations, vector<double>& values, vector<vector<string>>& queries) {
        unordered_map<string, string> baseof;
        unordered_map<string, unordered_map<string,double>> valuein;
        for(auto i=0u; i<equations.size(); ++i){
            auto numer = equations[i][0];
            auto denom = equations[i][1];
            if(baseof.count(numer)==0){
                if(baseof.count(denom)==0){ // both are new
                    baseof[numer] = denom;
                    baseof[denom] = denom;
                    valuein[denom][numer] = values[i];
                    valuein[denom][denom] = 1.0;
                }else{ // numer new, denom known
                    auto base = baseof[denom];
                    baseof[numer] = base;
                    valuein[base][numer] = values[i] * valuein[base][denom];
                }
            }else{
                if(baseof.count(denom)==0){ // numer known, denom new
                    auto base = baseof[numer];
                    baseof[denom] = base;
                    valuein[base][denom] = valuein[base][numer] / values[i];
                }else{ // both known
                    auto b_numer = baseof[numer];
                    auto b_denom = baseof[denom];
                    if(b_numer!=b_denom){ // set union
                        string mainbase, subbase;
                        double ratio;
                        if(valuein[b_numer].size() <= valuein[b_denom].size()){
                            mainbase = b_denom;
                            subbase = b_numer;
                            ratio = values[i] * valuein[mainbase][denom] / valuein[subbase][numer];
                        }else{
                            mainbase = b_numer;
                            subbase = b_denom;
                            ratio = valuein[mainbase][numer] / (values[i] * valuein[subbase][denom]);
                        }
                        for(auto [k,v]: valuein[subbase]){
                            valuein[mainbase][k] = v * ratio;
                            baseof[k] = mainbase;
                        }
                        valuein.erase(subbase);
                    }
                }
            }
        }
        vector<double> ans;
        for(auto j=0u; j<queries.size(); ++j){
            auto numer = queries[j][0];
            auto denom = queries[j][1];
            if(baseof.count(numer)==0 || baseof.count(denom)==0 || baseof[numer]!=baseof[denom]){
                ans.push_back(-1.0);
            }else{
                auto base = baseof[numer];
                ans.push_back(valuein[base][numer] / valuein[base][denom]);
            }
        }
        return ans;
    }
};

File: 399-evaluate-division.py
from collections import defaultdict
from typing import Dict, List

class Solution:
    def calcEquation(self, equations: List[List[str]], values: List[float], queries: List[List[str]]) -> List[float]:
        baseof: Dict[str, str] = dict()
        valuein: defaultdict[str, Dict[str, float]] = defaultdict(dict)
        for (numer, denom), v in zip(equations, values):
            if numer not in baseof:
                if denom not in baseof: # both are new
                    baseof[numer] = baseof[denom] = denom
                    valuein[denom][numer] = v
                    valuein[denom][denom] = 1.0
                else: # numer new, denom known
                    base = baseof[denom]
                    baseof[numer] = base
                    valuein[base][numer] = valuein[base][denom] * v
            else:
                if denom not in baseof: # numer known, denom new
                    base = baseof[numer]
                    baseof[denom] = base
                    valuein[base][denom] = valuein[base][numer] / v
                else: # both known
                    b_numer, b_denom = baseof[numer], baseof[denom]
                    if b_numer != b_denom: # set union
                        mainb, subb = str(), str()
                        ratio = float()
                        if len(valuein[b_numer]) <= len(valuein[b_denom]):
                            mainb, subb = b_denom, b_numer
                            ratio = valuein[mainb][denom] * v / valuein[subb][numer]
                        else:
                            mainb, subb = b_numer, b_denom
                            ratio = valuein[mainb][numer] / (v * valuein[subb][denom])
                        for k,val in valuein[subb].items():
                            valuein[mainb][k] = val * ratio
                            baseof[k] = mainb
                        valuein.pop(subb)
                    else: # check contradiction
                        if valuein[b_numer][numer] / valuein[b_numer][denom] != v:
                            return list()
        ans: List[float] = list()
        for numer, denom in queries:
            if (numer not in baseof) or (denom not in baseof) or (baseof[numer] != baseof[denom]):
                ans.append(-1.0)
            else:
                base = baseof[numer]
                ans.append(valuein[base][numer] / valuein[base][denom])
        return ans